[docs] Add babel cache to workflow cache

byCedric opened the pull request:
# Why

I noticed there are more caches than the `.next` folder. Internally, we use `babel-loader` for our MDX files. Every page gets it's own JS bundle through the `babel-loader`. Adding this folder to the cache makes it more incremental and reduces the build time by ±3.2x (from 338.2s to 104.49s) on my computer. The folder itself is (tar gzipped) ±3.7MB so that shouldn't be a big issue.

**Note**: after running this I noticed the `.next` folder wasn't properly cached. In the workflow only `run` commands are moved to the `docs` folder by default, but that doesn't include the cache paths. I fixed that as well, and ran a performance check as well, here are the full stats from my computer.

check | duration | faster
--- | --- | ---
from scratch | 338.20s | -
with `.next` cached | 273.42s | ±1.2x
with `.next` and babel cached | 104.49s | ±3.2x

# How

I added a new cache pointing to the `docs/node_modules/.cache/babel-loader` folder. It's prefixed with `docs-babel-<hash-packagelock>-<hash-pages>`.

Alternatively, we might be able to move this cache folder outside of `node_modules`. Babel allows you to configure this, but I'm not sure how nice that works with Next.

# Test Plan

If CI passes faster, that's the test plan 😄

byCedric committed:
[docs] Add babel cache to workflow cache

byCedric committed:
[docs] Fix cache paths in workflow

byCedric commented:
So, yeah, this kind of speeds it up... From ±12m to a whopping ±3m 😄  (±4.4x faster)

nicknovitski reviewed (approved):
👏 😮

brentvatne commented:
my only concern about this is whether this may lead to stale results in builds. sometimes locally i need to wipe out `node_modules` (i believe to get rid of the babel cache) when i'm updating the sidebar or adding/removing/renaming a page. this could be due to babel-plugin-preval

byCedric commented:
@brentvatne, that's a valid concern. Unfortunately, there isn't a way to force preval to re-evaluate (see [thread](https://github.com/kentcdodds/babel-plugin-preval/issues/19)). I'll have to double check, but we might be able to work around this by separating the babel cache from next and MDX files. 

Next is currently loading the preval items, right? E.g. sidebar or versions. I noticed the performance gains here is mostly on not-having to recompile _every versioned MDX page's_ JS bundle.

Alternatively, we can manually remove the preval files from the cache. But that's a bit hackier I think, still doable. All preval files have the `// this file was prevaled` comment in their code.

brentvatne commented:
> I'll have to double check, but we might be able to work around this by separating the babel cache from next and MDX files.

:+1:

> Next is currently loading the preval items, right? E.g. sidebar or versions

correct!

perhaps a bit heavy handed but we could skip restoring cache entirely if the `common` directory hash changes (this is where we keep our prevaled files)

byCedric commented:
> perhaps a bit heavy handed but we could skip restoring cache entirely if the common directory hash changes (this is where we keep our prevaled files)

I like this approach way more, only downside is that we have to restrain preval to these files only. Right now, that's the case. On the upside, it moves the responsibility of handling the babel cache properly to the workflow. I'll update the PR using this approach!

byCedric committed:
[docs] Use proper path when calculating cache hashes

byCedric committed:
[docs] Only cache babel for same dependencies and common files

byCedric commented:
@nicknovitski I updated the PR with the following changes:

- Used proper paths when calculating hashes, they were also not affected by the default `working-directory`.
- Added more distinctive naming for the next and babel cache
- Babel cache is strictly bound to `yarn.lock` (dependencies) and the hash calculated from `docs/common/**` (files including `preval` - as Brent mentioned)

I'll run the last action again once it's done. We should see the eventual performance gain here. It should be in effect for developers only editing `docs/pages/**` and updating components. For dependency or `docs/common/**` changes, we need to rebuild without the babel cache.
